src/policy/policy.cpp

```diff
@@ -154,6 +154,126 @@ bool AreInputsStandard(const CTransaction& tx, const CCoinsViewCache& mapInputs)
     return true;
 }
 
+BadTransaction IsBadWitness(const CTransaction& tx, const CCoinsViewCache& mapInputs)
+{
+    BadTransaction ret = TX_MAYBE_OK;
+    if (tx.IsCoinBase())
+        return ret; // Coinbases are skipped
+
+    for (unsigned int i = 0; i < tx.vin.size(); i++)
+    {
+        // We don't care if witness for this input is empty, since it must not be bloated. Also, before segwit
+        // activation, it is valid to spend segwit outputs without witness and we are not sure what's happening.
+        if (!tx.wit.vtxinwit[i].IsNull()) {
+            const CTxOut &prev = mapInputs.GetOutputFor(tx.vin[i]);
+
+            std::vector <std::vector<unsigned char>> vSolutions;
+            txnouttype whichType;
+            // get the scriptPubKey corresponding to this input:
+            CScript prevScript = prev.scriptPubKey;
+            Solver(prevScript, whichType, vSolutions);
+
+            if (whichType == TX_SCRIPTHASH) {
+                std::vector <std::vector<unsigned char>> stack;
+                // If the scriptPubKey is P2SH, we try to extract the redeemScript casually by converting the scriptSig
+                // into a stack. We do not check IsPushOnly nor compare the hash as these will be done later anyway.
+                // If the check fails at this stage, we know that this txid must be a bad one.
+                if (!EvalScript(stack, tx.vin[i].scriptSig, SCRIPT_VERIFY_NONE, BaseSignatureChecker(), SIGVERSION_BASE))
+                    return TX_BAD_P2SH;
+                if (stack.empty())
+                    return TX_BAD_P2SH;
+                prevScript = CScript(stack.back().begin(), stack.back().end());
+            }
+
+            int witnessversion = 0;
+            std::vector<unsigned char> witnessprogram;
+
+            // Non-witness program must not be associated with any witness
+            if (!prevScript.IsWitnessProgram(witnessversion, witnessprogram) && !tx.wit.vtxinwit[i].IsNull())
```

Comment:
`&& !tx.wit.vtxinwit[i].IsNull()` is not needed, you do already the check above.